inky/impression.go

```diff
@@ -0,0 +1,414 @@
+// Copyright 2023 The Periph Authors. All rights reserved.
+// Use of this source code is governed under the Apache License, Version 2.0
+// that can be found in the LICENSE file.
+
+package inky
+
+import (
+	"encoding/binary"
+	"fmt"
+	"image"
+	"image/color"
+	"image/draw"
+	"log"
+	"time"
+
+	"periph.io/x/conn/v3"
+	"periph.io/x/conn/v3/display"
+	"periph.io/x/conn/v3/gpio"
+	"periph.io/x/conn/v3/physic"
+	"periph.io/x/conn/v3/spi"
+)
+
+var _ display.Drawer = &DevImpression{}
+var _ conn.Resource = &DevImpression{}
+var _ draw.Image = &DevImpression{}
+
+var (
+	// For more: https://github.com/pimoroni/inky/issues/115#issuecomment-887453065
+	dsc = []color.NRGBA{
+		{0, 0, 0, 0},         // Black
+		{255, 255, 255, 255}, // White
+		{0, 255, 0, 255},     // Green
+		{0, 0, 255, 255},     // Blue
+		{255, 0, 0, 255},     // Red
+		{255, 255, 0, 255},   // Yellow
+		{255, 140, 0, 255},   // Orange
+		{255, 255, 255, 255},
+	}
+
+	sc = []color.NRGBA{
+		{57, 48, 57, 0},      // Black
+		{255, 255, 255, 255}, // White
+		{58, 91, 70, 255},    // Green
+		{61, 59, 94, 255},    // Blue
+		{156, 72, 75, 255},   // Red
+		{208, 190, 71, 255},  // Yellow
+		{177, 106, 73, 255},  // Orange
+		{255, 255, 255, 255},
+	}
+)
+
+const (
+	UC8159PSR   = 0x00
+	UC8159PWR   = 0x01
+	UC8159POF   = 0x02
+	UC8159PFS   = 0x03
+	UC8159PON   = 0x04
+	UC8159BTST  = 0x06
+	UC8159DSLP  = 0x07
+	UC8159DTM1  = 0x10
+	UC8159DSP   = 0x11
+	UC8159DRF   = 0x12
+	UC8159IPC   = 0x13
+	UC8159PLL   = 0x30
+	UC8159TSC   = 0x40
+	UC8159TSE   = 0x41
+	UC8159TSW   = 0x42
+	UC8159TSR   = 0x43
+	UC8159CDI   = 0x50
+	UC8159LPD   = 0x51
+	UC8159TCON  = 0x60
+	UC8159TRES  = 0x61
+	UC8159DAM   = 0x65
+	UC8159REV   = 0x70
+	UC8159FLG   = 0x71
+	UC8159AMV   = 0x80
+	UC8159VV    = 0x81
+	UC8159VDCS  = 0x82
+	UC8159PWS   = 0xE3
+	UC8159TSSET = 0xE5
+)
+
+// DevImpression is a handle to an Inky Impression.
+type DevImpression struct {
+	*Dev
+
+	// Color Palette used to convert images to the 7 color.
+	Palette color.Palette
+	// Representation of the pixels.
+	Pix []uint8
+
+	// Saturation level used by the color palette.
+	saturation float64
```

Comment:
Historically I shied away from floating points in periph because early Raspberry Pis didn't have a FPU. This is becoming more rare nowadays.

That said using float64 gives the false impression to the user that there is more precision that there is in reality. I'd personally probably go with a uint8.

Reply:
Done.